charms/slurmctld/src/interface_influxdb.py

```diff
@@ -0,0 +1,257 @@
+#!/usr/bin/env python3
+"""Influxdb interface."""
+import json
+import logging
+
+import influxdb
+import requests
+from influxdb.exceptions import InfluxDBClientError
+from ops import (
+    EventBase,
+    EventSource,
+    Object,
+    ObjectEvents,
+    RelationBrokenEvent,
+    RelationJoinedEvent,
+)
+from utils import generate_random_string
+
+logger = logging.getLogger()
+
+
+class InfluxDBAvailableEvent(EventBase):
+    """InfluxDBAvailable event."""
+
+    def __init__(
+        self,
+        handle,
+        influxdb_host,
+        influxdb_user,
+        influxdb_pass,
+        influxdb_database,
+        influxdb_policy,
+    ):
+        super().__init__(handle)
+
+        self.influxdb_host = influxdb_host
+        self.influxdb_user = influxdb_user
+        self.influxdb_pass = influxdb_pass
+        self.influxdb_database = influxdb_database
+        self.influxdb_policy = influxdb_policy
+
+    def snapshot(self):
+        """Snapshot the event data."""
+        return {
+            "influxdb_host": self.influxdb_host,
+            "influxdb_user": self.influxdb_user,
+            "influxdb_pass": self.influxdb_pass,
+            "influxdb_database": self.influxdb_database,
+            "influxdb_policy": self.influxdb_policy,
+        }
+
+    def restore(self, snapshot):
+        """Restore the snapshot of the event data."""
+        self.influxdb_host = snapshot.get("influxdb_host")
+        self.influxdb_user = snapshot.get("influxdb_user")
+        self.influxdb_pass = snapshot.get("influxdb_pass")
+        self.influxdb_database = snapshot.get("influxdb_database")
+        self.influxdb_policy = snapshot.get("influxdb_policy")
+
+
+class InfluxDBUnavailableEvent(EventBase):
+    """InfluxDBUnavailable event."""
+
+
+class InfluxDBEvents(ObjectEvents):
+    """InfluxDBEvents."""
+
+    influxdb_available = EventSource(InfluxDBAvailableEvent)
+    influxdb_unavailable = EventSource(InfluxDBUnavailableEvent)
+
+
+class InfluxDB(Object):
```

Comment:
[question]: What's the difference between the application-level InfuxDB relation databag and the unit-level databag? What does the application leader provide that the units don't?

Reply:
Using the application databag gives us the persistence of the admin_info in the case slurmctld needs to go away and be replaced by a new unit.